### PR TITLE
[R4R]-[fee]feat: fee collection with tokenRatio

### PR DESCRIPTION
**Pull request by Tri-stone**
Core changes:
- get `tokenRatio` from [GasPriceOracle](https://github.com/mantlenetworkio/mantle-v2/blob/main/packages/contracts-bedrock/contracts/L2/GasPriceOracle.sol) 
- when estimateGas and executing a tx, it will involve tokenRatio to calculate `L1 Rollup Fee` & `L2 Execution Fee`
    - L2 Execution Fee = L2 gasPrice * L2 gasUsed * tokenRatio
    - L1 Rollup Fee = (rollupDataGas + overhead) * L1BaseFee * scalar
        - L1BaseFee = L1 gasPrice * tokenRatio
    - L2 gasUsed' =  (L2 Execution Fee + L1 Rollup Fee) / L2 gasPrice
        - L2 gasUsed' will be the final gasLimit

Releated PR:
- https://github.com/mantlenetworkio/mantle-v2/pull/41

### CODE REVIEW

File: accounts/abi/bind/backends/simulated.go [hunk 1]
```diff
@@ -739,6 +739,7 @@ func (b *SimulatedBackend) callContract(ctx context.Context, call ethereum.CallM
 	txContext := core.NewEVMTxContext(msg)
 	evmContext := core.NewEVMBlockContext(header, b.blockchain, nil, b.config, stateDB)
 	vmEnv := vm.NewEVM(evmContext, txContext, stateDB, b.config, vm.Config{NoBaseFee: true})
+	log.Info("SimulatedBackend")
```
Comment (HaoyangLiu) [inline accounts/abi/bind/backends/simulated.go:742]: Is this a necessary log?

File: core/blockchain.go [hunk 2]
```diff
@@ -151,6 +151,10 @@ var defaultCacheConfig = &CacheConfig{
 	SnapshotWait:   true,
 }
 
+var (
+	DefaultMantleBlockGasLimit = uint64(1125899906842624)
```
Comment (HaoyangLiu) [inline core/blockchain.go:155]: Use const value? `1125899906842624 ` -> `0x4000000000000`

File: core/genesis.go [hunk 3]
```diff
@@ -584,10 +584,11 @@ func DefaultGoerliGenesisBlock() *Genesis {
...
 	return &Genesis{
-		Config:     params.SepoliaChainConfig,
-		Nonce:      0,
-		ExtraData:  []byte("Sepolia, Athens, Attica, Greece!"),
-		GasLimit:   0x1c9c380,
+		Config:    params.SepoliaChainConfig,
+		Nonce:     0,
+		ExtraData: []byte("Sepolia, Athens, Attica, Greece!"),
+		//GasLimit:   0x1c9c380,
+		GasLimit:   0x4000000000000,
```
Comment (HaoyangLiu) [inline core/genesis.go:591]: `0x4000000000000` -> `DefaultMantleBlockGasLimit`

File: accounts/abi/bind/backends/simulated.go (hunk 1 above)
Comment (Tri-stone) [reply accounts/abi/bind/backends/simulated.go:742]: Yes, will delete it.

Comment (Tri-stone) [reply accounts/abi/bind/backends/simulated.go:742]: yes, will delete it.

File: core/blockchain.go (hunk 2 above)
Comment (Tri-stone) [reply core/blockchain.go]: modified

File: core/genesis.go (hunk 3 above)
Comment (Tri-stone) [reply core/genesis.go]: not necessary to set `GasLimit` here, deleted it.

**Review by byteflyfunny — APPROVED** [review]

**Review by byteflyfunny — APPROVED** [review]

File: core/vm/runtime/runtime.go [hunk 4]
```diff
@@ -74,7 +74,9 @@ func setDefaults(cfg *Config) {
 		cfg.Difficulty = new(big.Int)
 	}
 	if cfg.GasLimit == 0 {
-		cfg.GasLimit = math.MaxUint64
+		// todo: leo
+		// cfg.GasLimit = math.MaxUint64
```
Comment (HaoyangLiu) [inline core/vm/runtime/runtime.go:78]: Please resolve `todo`

File: core/state_transition.go [hunk 5]
```diff
@@ -128,6 +128,15 @@ func toWordSize(size uint64) uint64 {
 	return (size + 31) / 32
 }
 
+type RunMode uint8
+
+const (
+	CommitMode RunMode = iota
```
Comment (HaoyangLiu) [inline core/state_transition.go:134]: Please assign `CommitMode` to `RunMode` in function `TransactionToMessage`. Not a bug. It is strange that `CommitMode` is never used.

File: core/types/receipt.go [hunk 6]
```diff
@@ -499,7 +531,7 @@ func (rs Receipts) DeriveFields(config *params.ChainConfig, hash common.Hash, nu
 			// Deriving the signer is expensive, only do if it's actually needed
 			from, _ := Sender(signer, txs[i])
 			nonce := txs[i].Nonce()
-			if rs[i].DepositNonce != nil {
+			if txs[i].IsDepositTx() && rs[i].DepositNonce != nil {
```
Comment (abelliumt) [inline core/types/receipt.go:534]: `txs[i].IsDepositTx()` seems to be not necessary.

File: core/state_transition.go [hunk 7]
```diff
@@ -549,9 +613,10 @@ func (st *StateTransition) innerTransitionDb() (*ExecutionResult, error) {
 	// Note optimismConfig will not be nil if rules.IsOptimismBedrock is true
 	if optimismConfig := st.evm.ChainConfig().Optimism; optimismConfig != nil && rules.IsOptimismBedrock {
 		st.state.AddBalance(params.OptimismBaseFeeRecipient, new(big.Int).Mul(new(big.Int).SetUint64(st.gasUsed()), st.evm.Context.BaseFee))
-		if cost := st.evm.Context.L1CostFunc(st.evm.Context.BlockNumber.Uint64(), st.evm.Context.Time, st.msg.RollupDataGas, st.msg.IsDepositTx); cost != nil {
-			st.state.AddBalance(params.OptimismL1FeeRecipient, cost)
-		}
+		// leo: can not collect l1 fee here again, all l1 fee has been collected by CoinBase & OptimismBaseFeeRecipient
+		//if cost := st.evm.Context.L1CostFunc(st.evm.Context.BlockNumber.Uint64(), st.evm.Context.Time, st.msg.RollupDataGas, st.msg.IsDepositTx); cost != nil {
+		//	st.state.AddBalance(params.OptimismL1FeeRecipient, cost)
+		//}
```
Comment (abelliumt) [inline core/state_transition.go:619]: `params.OptimismBaseFeeRecipient` is useless now. So where will the collected l1 fee go now?

**Comment by abelliumt** [comment]
No other comment

File: core/vm/runtime/runtime.go (hunk 4 above)
Comment (Tri-stone) [reply core/vm/runtime/runtime.go:78]: resolved

File: core/state_transition.go (hunk 5 above)
Comment (Tri-stone) [reply core/state_transition.go:134]: fixed, set default `RunMode` as `CommitMode`

File: core/types/receipt.go (hunk 6 above)
Comment (Tri-stone) [reply core/types/receipt.go:534]: We changes the way to store `DepositNonce`, so the `DepositNonce` will not be nil and the default value will be 0 for non-depositTxs, so add `txs[i].IsDepositTx()` to determine the tx type.

**Review by abelliumt — APPROVED** [review]

**Review by byteflyfunny — APPROVED** [review]

**Pull request merged**